### DISCUSSION TITLE
Raspberry Pi High Quality Camera

### DISCUSSION THREAD

**Discussion by Ayanhb** [raspberrypi/picamera2#1002] [Q&A]
So I am having trouble with .DNG format. When I do it in the terminal I get some decent images (monochromed but 253 x 190 pixels). However, when I go into python (Visual Studio Code or Thonny), the .DNG image is just these white and grey boxes (4056 x 3040 Pixels)... does anyone know what I am doing wrong here?? The plan is for me to take both JPG and DNG, and then with the DNG convert it to a Tiff format. Then convert the Tiff to grayscale. And then from there I am able to calculate for the Albedo. PLEASE HELP

**Comment by Ayanhb** [comment]
<img width="535" alt="Screen Shot 2024-04-04 at 8 31 21 PM" src="https://github.com/raspberrypi/picamera2/assets/135760082/9575bd81-9f52-41da-9095-1898a0b4daeb">

**Comment by Ayanhb** [comment]
Here is when I ran it in the terminal OS and it is both JPG and the DNG file compared to the previous picture (where it was in Visual Studio Code)

![IMG_1619 2](https://github.com/raspberrypi/picamera2/assets/135760082/714232d7-5953-4025-ba63-39391ac0eace)
![IMG_1620](https://github.com/raspberrypi/picamera2/assets/135760082/3cc9ef2b-2dac-4dc8-817b-3e8707cecf92)

**Comment by davidplowman** [comment]
Hi, thanks for the question.

I think the problem may simply be that the image viewer does not handle DNG files correctly. It discovers that it doesn't know how to decode the image, so you end up with the "background", which is the white/grey checkerboard pattern, showing through. (Viewing a completely transparent image file usually looks the same.)

When you use our command line applications, these save a small greyscale thumbnail. DNG files sometimes contain thumbnails like this because they know that some viewer applications will only understand and render the thumbnail. But the Python DNG library that we use does not save a thumbnail (hence why the image doesn't show up).

You could convert the DNG files to regular image files to view them. The easiest way is probably to use dcraw:
```
dcraw -w <dng-filename>
```
which will create a `<dng-filename>.ppm` file that all image viewers should understand. If you don't have dcraw on your system, I think it's just `sudo apt install dcraw`.

**Reply by Ayanhb** [reply]
I actually just figured that out a little while ago. Using dcraw it actually created a .ppm format and I was able to see the image. I am trying to figure out how to get it to Tiff from dcraw since I read that is possible (haven't got that yet). And then I have to convert that to grayscale as well. 

Issue that I ran into is that I wrote a code to read the pixel data for the file but it keeps freezing my raspberry pi os since its so much data... need to find out something to see all of that because I need this data to calculate the albedo! Any recommendations or codes I can look at for a better understanding for these issues?

**Comment by davidplowman** [comment]
I would consider doing the processing directly in Python using Picamera2. Note that dcraw turns the raw image into a processed image, so if that's what you want you may as well use the processed image directly from Picamera2. (I assume you're familiar with the fact that a DNG file contains a Bayer pattern which needs a certain amount of processing before it is "viewable".)

With Picamera2 you can grab a raw and a processed image simultaneously. You can request the processed image in either RGB or YUV420 (where the Y gives you greyscale directly). You have numpy and OpenCV  to do any processing that you need, and I believe OpenCV supports writing to TIFF files too.

**Reply by Ayanhb** [reply]
So whenever I did do a raw and a processed image (which is the example code in this GitHub), the DNG file is always that black and white boxes. But I am not sure if I did the code correctly for the processed image. But also I need the uncompressed format and then convert to grayscale to be able to get all the data and calculate for the albedo (at least that is what my advisor told me). 

Thats why I tried the dcraw. I did that process with the dng and jpg, then I used dcraw to make it into .ppm, and then from there tried to read the pixel data but that's when it crashed and I stopped for the night. Thinking I am just not understanding what you are saying or something like that haha.

**Comment by davidplowman** [comment]
I suspect there was nothing wrong with your DNG file, it's just that the image viewer cannot handle it properly. (As I said, try dcraw on it to be sure!)

I also suspect a little bit that you don't want the raw file at all, because that's a Bayer pattern and not a regular image. It needs, usually, lots of processing to be usable. Is it perhaps the case that you want an *uncompressed* image, and not a raw image? (Unfortunately, in the video community they sometimes use "raw" to mean "uncompressed", but in the image sensor community, and therefore in libcamera and Picamera2, "raw" means "Bayer".)

**Reply by Ayanhb** [reply]
Yeah it was just the image viewer because dcraw worked fine for it so at least that is behind me haha.

But I did not know that raw for the Picamera2 meant bayer... because I need uncompressed not compressed. But then I think I am fine because I haven't been using the raw.py, instead the dng.py from the GitHub. Which then when I used exiftool, it says it is uncompressed. But now after reading all this I am not to sure to be honest...

**Comment by davidplowman** [comment]
So my advice would be to configure the camera to give you "YUV420", and then use `capture_array` to capture the image. Just take the first `image_height` rows from this, which will give you a greyscale image (the rows that you are throwing away are the colour information). Use cv2 directly to save this as png (which is uncompressed) or in any other of its supported formats.

**Reply by Ayanhb** [reply]
Basically something like this:

import cv2
import picamera2 as Picamera2

# Create a Picamera2 instance
picamera2 = Picamera2()

# Configure the camera to capture in YUV420 format
config = picamera2.create_video_configuration(main={"size": (4056, 3040), "format": 'YUV420'},
                                              loopback=False)
picamera2.configure(config)
picamera2.start()

# Capture an image from the camera
array = picamera2.capture_array()

# Stop the camera
picamera2.stop()

# Extract the greyscale image from the YUV420 array
greyscale_image = array[:image_height, :]

# Save the greyscale image as a PNG image
cv2.imwrite('greyscale_image.png', greyscale_image)

**Reply by davidplowman** [reply]
Yes, exactly. Just a couple of very small things:

* I would consider `create_still_configuration` instead of `create_video_configuration` because it will allocate fewer buffers by default, which may be an issue on non-Pi 5s. Also you'll get slightly better image quality on a non-Pi 5.
* Optionally have a short `time.sleep(0.5)` before the capture to let the AGC and AWB algorithms settle. This may matter less if you're saving a greyscale image (as the AWB isn't important!).

**Reply by Ayanhb** [reply]
Ah okay, that was my bad. It was late last night and didn't double check the configuration! I modified it to be like this:

import cv2
import picamera2 as Picamera2

# Create a Picamera2 instance
picamera2 = Picamera2()

# Configure the camera to capture in YUV420 format
config = picamera2.create_still_configuration(main={"size": (4056, 3040), "format": 'YUV420'},
                                              loopback=False)
picamera2.configure(config)

# Optionally add a short sleep to let the AGC and AWB algorithms settle
import time
time.sleep(0.5)

# Capture an image from the camera
array = picamera2.capture_array()

# Stop the camera
picamera2.stop()

# Extract the greyscale image from the YUV420 array
greyscale_image = array[:image_height, :]

# Save the greyscale image as a PNG image
cv2.imwrite('greyscale_image.png', greyscale_image)

**Reply by davidplowman** [reply]
Yes, give that a try. (Tip: to include a block of code in the discussion, use a line containing just three backquotes at the start and end which will make it easier to read and understand.)
